sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala

```diff
@@ -60,28 +60,42 @@ object ExternalCatalogUtils {
       Array(' ', '<', '>', '|').foreach(bitSet.set(_))
     }
 
-    bitSet
+    (bitSet, bitSet.size)
   }
 
-  def needsEscaping(c: Char): Boolean = {
-    c < charToEscape.size() && charToEscape.get(c)
+  private final val HEX_CHARS = "0123456789ABCDEF".toCharArray
+
+  @inline final def needsEscaping(c: Char): Boolean = {
+    c < sizeOfCharToEscape && charToEscape.get(c)
   }
 
   def escapePathName(path: String): String = {
-    val builder = new StringBuilder()
-    path.foreach { c =>
-      if (needsEscaping(c)) {
-        builder.append('%')
-        builder.append(f"${c.asInstanceOf[Int]}%02X")
-      } else {
-        builder.append(c)
+    if (path == null || path.isEmpty) {
```

Comment:
Should we directly return if `path` is a blank string? For example,  `"              "`

maybe we should use `lang3.StringUtils.isBlank(path)`

Reply:
Nope, we can't do that.

```
scala> org.apache.commons.lang3.StringUtils.isBlank("\n")
val res2: Boolean = true

scala> org.apache.spark.sql.catalyst.catalog.ExternalCatalogUtils.escapePathName("\n")
val res3: String = %0A
```

Reply:
Got